nni/compression/pytorch/speedup/compress_modules.py

```diff
@@ -3,196 +3,381 @@
+    new_norm.bias.data = torch.index_select(norm.bias.data, 0, remained_in)
+
+    new_norm.running_mean.data = torch.index_select(
+        norm.running_mean.data, 0, remained_in)
+    new_norm.running_var.data = torch.index_select(
+        norm.running_var.data, 0, remained_in)
+    return new_norm
+
+
+def replace_batchnorm2d(norm, auto_infer):
     """
     Parameters
     ----------
     norm : torch.nn.BatchNorm2d
         The batchnorm module to be replace
-    mask : ModuleMasks
-        The masks of this module
+    auto_infer : AutoMaskInference
+        The auto mask inference object that contains the input,
+        parameter and output masks.
 
     Returns
     -------
     torch.nn.BatchNorm2d
         The new batchnorm module
     """
-    assert isinstance(mask, ModuleMasks)
-    assert 'weight' in mask.param_masks and 'bias' in mask.param_masks
-    index = mask.param_masks['weight'].mask_index[0]
-    num_features = index.size()[0]
-    _logger.debug("replace batchnorm2d with num_features: %d", num_features)
+    assert isinstance(norm, nn.BatchNorm2d)
+    in_mask = auto_infer.in_masks[0]
+    output_mask = auto_infer.output_mask
+
+    # N, C, H, W
+    _, remained_in = convert_to_coarse_mask(in_mask, 1)
+    _, remained_out = convert_to_coarse_mask(output_mask, 1)
+    assert remained_in.size(0) == remained_out.size(0)
+
+    num_features = remained_in.size(0)
+    _logger.info("replace batchnorm2d with num_features: %d", num_features)
     new_norm = torch.nn.BatchNorm2d(num_features=num_features,
                                     eps=norm.eps,
                                     momentum=norm.momentum,
                                     affine=norm.affine,
                                     track_running_stats=norm.track_running_stats)
     # assign weights
-    new_norm.weight.data = torch.index_select(norm.weight.data, 0, index)
-    new_norm.bias.data = torch.index_select(norm.bias.data, 0, index)
-    if norm.track_running_stats:
-        new_norm.running_mean.data = torch.index_select(
-            norm.running_mean.data, 0, index)
-        new_norm.running_var.data = torch.index_select(
-            norm.running_var.data, 0, index)
+    new_norm.weight.data = torch.index_select(norm.weight.data, 0, remained_in)
+    new_norm.bias.data = torch.index_select(norm.bias.data, 0, remained_in)
+
+    new_norm.running_mean.data = torch.index_select(
+        norm.running_mean.data, 0, remained_in)
+    new_norm.running_var.data = torch.index_select(
+        norm.running_var.data, 0, remained_in)
     return new_norm
 
 
-def replace_conv2d(conv, mask):
+class BiasModule(nn.Module):
+    def __init__(self, module, bias):
+        super(BiasModule, self).__init__()
+        self.ori_module = module
+        self.register_buffer('speedup_bias', bias)
+
+    def forward(self, x):
+        return self.ori_module(x)[:] + self.speedup_bias
+
+
+def replace_conv2d(conv, auto_infer):
```

Comment:
could you brief explain in high-level why the logic of replacing module increases a lot?

Reply:
Basically, I enhance the replace function of the conv2d, in the original version, it can remove the whole group of filter only when this conv is depth-wise. Currently, we can remove the whole group of filters once the filters in the same groups are masked out.